File: 7_kyu/Least_Larger.go
package kata

import "math"

// Returns the index of the smallest number that is larger than the element at the given index.
// Returns -1 if there is no such index.
func LeastLarger(a []int, i int) int {
	limit := a[i]
	// Initialise with maximal value.
	// If this value or a smalller value is in the array,
	// the the minIndex is set.
	min := math.MaxInt
	minIndex := -1
	for index, value := range a {
		// Use <= compare because value maxint must change the maxIndex.
		if limit < value && value <= min {
			// Found a new minimal value above limit
			min = value
			minIndex = index
		}
	}
	return minIndex
}

File: 7_kyu/Least_Larger.rs
// Returns index of the least item larger a[i].
// Returns None if no element exists, returns last min if more than one exists.
fn least_larger(a: &[i32], i: usize) -> Option<usize> {
    if i >= a.len() {
        return None;
    }
    // The value of the element is the lower limit (exlusive)
    let limit = a[i];
    // search minimal element above limit
    let mut min = i32::MAX;
    let mut min_index: Option<usize> = None;
    for (i, &x) in a.iter().enumerate() {
        // check lerr-equal (!) to catch a value with i32::MAX in the array.
        if x > limit && x <= min {
            min_index = Some(i);
            min = x;
        }
    }
    min_index
}
